### DISCUSSION TITLE
can Plot select a window's beginning and ending data?

### DISCUSSION THREAD

**Discussion by tx0c** [observablehq/plot#635] [Q&A] [answered]
I'm doing a graph like this https://observablehq.com/d/2d5429036206e1ae

the data is a monthly series data, to show a website's active users, as the column `month` `num_users` implies; also because the site's getting new users every month, and lost some as well, the `num_users_m3a` is showing how many matured users (age >= 3months) are still active;
```json
  0: Object {month: 2021-12-01, num_users: 10957, num_users_m3a: 4684}
  1: Object {month: 2021-11-01, num_users: 10529, num_users_m3a: 3974}
  2: Object {month: 2021-10-01, num_users: 10357, num_users_m3a: 5638}
  3: Object {month: 2021-09-01, num_users: 10226, num_users_m3a: 5554}
  4: Object {month: 2021-08-01, num_users: 9936, num_users_m3a: 5247}
```

here want to add a link between each 3 months pair, something like this but this does not work,

```js
      Plot.link(stats, {
        x1: d => d3.utcMonth(d.month, -3), x2: "month",
        y1: "num_users", // here actually want to select the prior 3months' num_users data
        y2: "num_users",
        stroke: "red",
      })
```

I tried `Plot.window()` but cannot find a proper reduce function, I guess the end result would be some sort of `Plot.select({ d1: the-begin, d2: the-end }, Plot.window({ k: 4, reduce: /* to output the window array of data instead of any current function mean/median/... */ }))`, the link will create a line between `month: 2021-08-01, num_users: 9936` toward `month: 2021-11-01, num_users_m3a: 3974`
```
  1: Object {month: 2021-11-01, num_users: 10529, num_users_m3a: 3974}
  4: Object {month: 2021-08-01, num_users: 9936, num_users_m3a: 5247}
```

**Comment by Fil** [comment, marked as answer]
you could try and use
~~~js
Plot.link(stats, Plot.map({
        y1: Plot.window({ k: 4, anchor: "end", reduce: d => d[0] }),
        x1: Plot.window({ k: 4, anchor: "end", reduce: d => d[0] }),
      }, {
        x1: "month",
        x2: "month",
        y1: "num_users_m3a",
        y2: "num_users",
        stroke: "red",
      }))
~~~
![untitled (1)](https://user-images.githubusercontent.com/7001/147778658-227aba4e-3e16-428e-b4f9-bd8b73e1d875.png)

**Reply by Fil** [reply]
or better with your midMonth helper:

~~~js
Plot.link(stats, Plot.map({
        x1: Plot.window({ k: 4, anchor: "end", reduce: d => midMonth(d[0]) }),
        x2: data => data.map(midMonth),
        y1: Plot.window({ k: 4, anchor: "end", reduce: d => d[0] })
      }, {
        x1: "month",
        x2: "month",
        y1: "num_users_m3a",
        y2: "num_users",
        stroke: "red",
      }))
~~~

![untitled (2)](https://user-images.githubusercontent.com/7001/147779312-dcbb3ca5-ed51-4c38-ac79-7cdf77991dbe.png)

**Reply by mbostock** [reply]
If you aren’t doing any fancy grouping or faceting, and your data is in chronological order, you can offset the values using array.slice like so:

```js
Plot.link(stats, {
  x1: stats.map(d => midMonth(d.month)).slice(3),
  y1: stats.map(d => d.num_users).slice(3),
  x2: d => midMonth(d.month),
  y2: d => d.num_users_m3a,
  stroke: "red",
})
```
<img width="1170" alt="Screen Shot 2021-12-30 at 12 28 49 PM" src="https://user-images.githubusercontent.com/230541/147786032-fb90ca1f-2e85-42dd-bf3b-2815adbcb378.png">

**Reply by tx0c** [reply]
> Plot.link(stats, Plot.map({

ok, it is that `Plot.map` doing the heavy lifting work which is hard to understand and I overlooked
